net/keepalived/files/usr/share/keepalived/scripts/rsync.sh

```diff
@@ -0,0 +1,153 @@
+#!/bin/sh
+
+. /lib/functions.sh
+
+SYNC_USER=keepalived-sync
+SYNC_USER_ID=60001
+SYNC_DIR=/usr/share/keepalived/rsync
+
+utc_timestamp() {
+	date -u +%s
+}
+
+update_last_sync_time() {
+	uci_revert_state keepalived "$1" last_sync_time
+	uci_set_state keepalived "$1" last_sync_time "$(utc_timestamp)"
+}
+
+update_last_sync_status() {
+	local cfg=$1
+	shift
+	local status="$@"
+	uci_revert_state keepalived "$cfg" last_sync_status
+	uci_set_state keepalived "$cfg" last_sync_status "$status"
+}
+
+ha_sync_send() {
+	local cfg=$1
+	local address ssh_user ssh_port sync_list sync_dir sync_file
+	local DIRS FILES
+
+	config_get address $cfg address
+	[ -z "$address" ] && return 0
+
+	config_get ssh_user $cfg ssh_user $SYNC_USER
+	config_get ssh_port $cfg ssh_port 22
+	config_get ssh_key $cfg ssh_key /root/.ssh/id_rsa
+	config_get sync_dir $cfg sync_dir $SYNC_DIR
+	config_get sync_list $cfg sync_list
+
+	for sync_file in $sync_list $(sysupgrade -l); do
+		[ -f "$sync_file" ] && {
+			dir="${sync_file%/*}"
+			list_contains FILES ${sync_file} || append FILES ${sync_file}
+		}
+		[ -d "$sync_file" ] && dir=${sync_file}
+		list_contains DIRS ${sync_dir}${dir} || append DIRS ${sync_dir}${dir}
+	done
+
+	changed_files=$(rsync --out-format='%n' --dry-run -a --relative $FILES -e "ssh -y -i $ssh_key -p $ssh_port" --rsync-path="sudo rsync" $ssh_user@$address:$sync_dir | wc -l)
+	if [ $? -ne 0 ]; then
+		update_last_sync_time "$cfg"
+		update_last_sync_status "$cfg" "Rsync Detection Failed"
+		return 0
+	elif [ $changed_files -le 0 ]; then
+		update_last_sync_time "$cfg"
+		update_last_sync_status "$cfg" "Up to Date"
+		return 0
+	fi
+
+	ssh -y -i $ssh_key -p $ssh_port $ssh_user@$address mkdir -m 755 -p $DIRS || {
+		update_last_sync_time "$cfg"
+		update_last_sync_status "SSH Connection Failed"
+		return 0
+	}
+
+	rsync -a --relative $FILES -e "ssh -y -i $ssh_key -p $ssh_port" --rsync-path="sudo rsync" $ssh_user@$address:$sync_dir || {
+		update_last_sync_time "$cfg"
+		update_last_sync_status "$cfg" "Rsync Transfer Failed"
+	}
+
+	update_last_sync_time "$cfg"
+	update_last_sync_status "$cfg" "Successful"
+}
+
+ha_sync_receive() {
```

Comment:
I like this solution :+1: 

I would define a [USERID](https://github.com/openwrt/packages/blob/6d49ad9e5c5a93014cc99f1c956eaa51b36ee95f/net/gnunet/Makefile#L64) in the Makefile to add a userid and groupid for keepalived. Make a suggestion of the numbers. I do not know what is already in use in the repostitories. I would prefer to use the group `keepalived` and user `keepalived` instead of `keepalived-sync`

Instead of generation the files in `/etc/sudoers.d/*` and `<user>/.ssh` we could use a [postinst](
https://github.com/openwrt/packages/search?q=%2Fpostinst&type=) script in the Makefile. We could also remove this files by usind [postrm](https://github.com/openwrt/packages/search?q=%2Fpostrm&type=)

Reply:
> I would define a [USERID](https://github.com/openwrt/packages/blob/6d49ad9e5c5a93014cc99f1c956eaa51b36ee95f/net/gnunet/Makefile#L64) in the Makefile to add a userid and groupid for keepalived. Make a suggestion of the numbers.

I don't think this is helping much because it creates a user with /bin/false shell and a home directory in /var/run and there is no option to pass custom values for home and shell to internal function `add_group_and_user` to handle it, 

So even if we set it in Makefile, we still need home directory creation, permission, and ownership to be set from the same code because even if we use this, It still needs to update in `/etc/passwd` for home and shell

Reply:
If that is the case with `USERID` then I would suggest that we move the creation/deletion of a user and the other stuff to the postinst/postrm scripts.
We don't have to do that every time on script calling.
I would only add/change the SSH key in the script.